src/coap.c

```diff
@@ -0,0 +1,311 @@
+/*
+ * Copyright (c) 2015 Cesanta Software Limited
+ * All rights reserved
+ * This software is dual-licensed: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation. For the terms of this
+ * license, see <http://www.gnu.org/licenses/>.
+ *
+ * You are free to use this software under the terms of the GNU General
+ * Public License, but WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * Alternatively, you can license this software under a commercial
+ * license, as set out in <http://cesanta.com/>.
+ */
+
+#ifndef NS_DISABLE_COAP
+
+/* 
+ * CoAP implementation.
+ *
+ * General CoAP message format:
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-
+ * |Ver| T | TKL | Code | Message ID | Token (if any, TKL bytes) ...
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-
+ * | Options (if any) ...            |1 1 1 1 1 1 1 1| Payload (if any) ...
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-
+ */
+
+/* TODO(alashkin): remove this #include after dev completion */
+#include <stdlib.h>
+
+#include "coap.h"
+
+/* Options memory management functions */
+void ns_coap_free_options(struct ns_coap_message *cm) {
+  while (cm->options != NULL) {
+    struct ns_coap_option *next = cm->options->next;
+    free(cm->options);
+    cm->options = next;
+  }
+}
+
+void ns_coap_add_option(struct ns_coap_message *cm,
+                        uint16_t number, char* value, size_t len) {
+  struct ns_coap_option *option =
+      (struct ns_coap_option *)NS_CALLOC(1, sizeof(*option));
+
+  option->number = number;
+  option->value.p = value;
+  option->value.len = len;
+
+  if (cm->options == NULL) {
+    cm->options = cm->_options_tail = option;
+  } else {
+    cm->_options_tail = cm->_options_tail->next = option;
```

Comment:
Have you meant message composing or parsing?
Composing: CoAP wants to have options ASC sorted. And here there are two options:
1. Just return an error, if client puts options in wrong way.
2. Resort provided options and put them to packet in right order.

At the moment p.1 is implemented. 
I think about adding sorting. But I was asked about minimal version, so, in any case implementation will have some kind of limitations.
I'm going to write design doc and describe limitations. After that we can discuss them all and choose something to implement.

Anyway, it will be in next PR.